packages/@vue/cli-service/lib/config/css.js

```diff
@@ -58,6 +59,19 @@ module.exports = (api, rootOptions) => {
     ]))
 
     if (!hasPostCSSConfig) {
+      // #6342
+      // NPM 6 may incorrectly hoist postcss 7 to the same level of autoprefixer
+      // So we have to run a preflight check to tell the users how to fix it
+      const autoprefixerDirectory = path.dirname(require.resolve('autoprefixer/package.json'))
+      const postcssPkg = loadModule('postcss/package.json', autoprefixerDirectory)
+      const postcssVersion = postcssPkg.version
+      if (!semver.satisfies(postcssVersion, '8.x')) {
+        throw new Error(
+          `The package manager has hoisted a wrong version of ${chalk.cyan('postcss')}, ` +
+          `please run ${chalk.cyan('npm i postcss@8 -D')} to fix it.`
+        )
+      }
+
```

Comment:
What if `hasPostCSSConfig` is `true`?

Reply:
That means the user is not likely using the built-in `autoprefixer` plugin.
In that case, they should take care of the peer dependency requirements of their own installed plugins.

Reply:
Projects created by Vue CLI < 4.1.0 may have a PostCSS configuration (which only adds `autoprefixer`) but no direct dependency installed. That's a bad practice. We can implement a migrator to help remove these configurations.